### PR TITLE
Notify Spar when a team is deleted

### PR DESCRIPTION
**Pull request by neongreen**
* [x] Delete tokens when the team is deleted
* [x] ~Delete tokens when the issuer is deleted?~
* [x] PRs to hegemony and cailleach about adding Spar to galley.yaml

### CODE REVIEW

**Comment by neongreen** [comment]
@fisx What other things should I do in Spar when a team is deleted? Should I delete relevant entries from `team_idp`?

**Comment by fisx** [comment]
> @fisx What other things should I do in Spar when a team is deleted? Should I delete relevant entries from `team_idp`?

tables to be cleaned up from looking at `services/spar/schema/src/V*.hs`: `user`, `idp`, `issuer_idp`, `team_idp`, `team_provisioning_by_team`, `team_provisioning_by_token`.

it should be straight-forward to decide which rows are dangling and need to removed, no?

the other tables should be harmless: bind cookies, authn_requests, assertions, verdict formats are all only needed during or after an authentication response has been validated, which won't be successful without the issuer mapping to an idp id.

not sure this answers your question?

**Comment by fisx** [comment]
> @fisx Please look at the last commit.

looks good!  i'll review the rest now.

**Comment by neongreen** [comment]
@fisx The test has been written by me and Tiago

**Comment by neongreen** [comment]
Me and Tiago wrote the test

File: services/spar/src/Spar/API.hs [hunk 1]
```diff
@@ -249,15 +249,5 @@ internalStatus = pure NoContent
...
-    Data.deleteTeamScimTokens team
-    -- Since IdPs are not shared between teams, we can look at the set of IdPs
-    -- used by the team, and remove everything related to those IdPs, too.
-    idps <- Data.getIdPConfigsByTeam team
-    for_ idps $ \idp -> do
-      let idpid = idp ^. SAML.idpId
-          issuer = idp ^. SAML.idpMetadata . SAML.edIssuer
-      Data.deleteIdPConfig idpid issuer team
-      Data.deleteUsersByIssuer issuer
+  wrapMonadClient $ Data.deleteTeam team
```
Comment (fisx) [inline services/spar/src/Spar/API.hs]: good idea!

File: services/spar/test-integration/Test/Spar/SCIMSpec.hs [hunk 2]
```diff
@@ -11,11 +11,11 @@
 
 module Test.Spar.SCIMSpec where
 
+import Imports
 import Bilge
 import Bilge.Assert
 import Control.Lens
 import Data.ByteString.Conversion
-import Imports
```
Comment (fisx) [inline services/spar/test-integration/Test/Spar/SCIMSpec.hs]: sorry...  :-/

File: services/spar/test-integration/Test/Spar/DataSpec.hs [hunk 3]
```diff
@@ -203,20 +210,44 @@ testSPStoreID store unstore isalive = do
...
+    let getUid = read . unpack . SCIM.Common.id . SCIM.Meta.thing
+    ssoid1 <- getSsoidViaSelf (getUid storedUser1)
+    ssoid2 <- getSsoidViaSelf (getUid storedUser2)
+    -- Delete the team
+    runSparCass $ Data.deleteTeam tid
+    -- See that everything got cleaned up.
+    --
+    -- The token from 'team_provisioning_by_token':
+    do tokenInfo <- runSparCass $ Data.lookupScimToken tok
+       liftIO $ tokenInfo `shouldBe` Nothing
```
Comment (fisx) [inline services/spar/test-integration/Test/Spar/DataSpec.hs]: [mumbling to myself] there is `shouldReturn`, which would make this more concise *and* arguably more readable, but it's in IO and not in MonadIO, so I'm not sure how to rewrite this.  i really should write mutants of all the `should*` functions in MonadIO...

File: services/spar/test-integration/Test/Spar/DataSpec.hs [hunk 4]
```diff
@@ -203,20 +210,44 @@ testSPStoreID store unstore isalive = do
...
+       mbUser2 <- runSparCass $ Data.getUser uref2
+       liftIO $ mbUser2 `shouldBe` Nothing
+    -- The config from 'idp':
+    do mbIdp <- runSparCass $ Data.getIdPConfig (idp ^. SAML.idpId)
+       liftIO $ mbIdp `shouldBe` Nothing
+    -- The config from 'issuer_idp':
+    do let issuer = idp ^. SAML.idpMetadata . SAML.edIssuer
+       mbIdp <- runSparCass $ Data.getIdPIdByIssuer issuer
+       liftIO $ mbIdp `shouldBe` Nothing
+    do idps <- runSparCass $ Data.getIdPConfigsByTeam tid
```
Comment (fisx) [inline services/spar/test-integration/Test/Spar/DataSpec.hs]: this one doesn't have a comment line.  :-)

File: services/spar/test-integration/Test/Spar/DataSpec.hs [hunk 5]
```diff
@@ -203,20 +210,44 @@ testSPStoreID store unstore isalive = do
         isit <- runSparCassWithEnv $ isalive xid
         liftIO $ isit `shouldBe` True
 
-    context "after TTL" $ do
```
Comment (fisx) [inline services/spar/test-integration/Test/Spar/DataSpec.hs]: why did these two tests go away?  can i have them back?

File: services/spar/test-integration/Test/Spar/DataSpec.hs (hunk 3 above)
Comment (neongreen) [reply services/spar/test-integration/Test/Spar/DataSpec.hs]: `should` functions should absolutely be in `MonadIO` and I've also been mumbling to myself about this for quite some time.

We can use http://hackage.haskell.org/package/hspec-expectations-lifted-0.10.0/docs/Test-Hspec-Expectations-Lifted.html but I'd rather have a homegrown library which would also contain some other common expectations that I end up reimplementing from time to time.

File: services/spar/test-integration/Test/Spar/DataSpec.hs (hunk 5 above)
Comment (neongreen) [reply services/spar/test-integration/Test/Spar/DataSpec.hs]: I have no idea :/ I'll get them back.

**Comment by neongreen** [comment]
Fixed everything. Sorry for the force-push though

**Pull request merged**